### PR TITLE
[MNG-8210] Replace Maven "module" term by "subproject"

### PR DESCRIPTION
**Pull request by gnodet**
Based on @desruisseaux initial commit to rename subproject #1649.
This PR adds `subprojects` to 4.1.0 POM, deprecate `modules`, and add a discovery mechanism which kicks in when:
  * pomVersion == 4.1.0
  * packaging == pom
  * no modules or subprojects are specified

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/project/MavenProject.java [hunk 1]
```diff
@@ -708,6 +708,9 @@ public List<Plugin> getBuildPlugins() {
     }
 
     public List<String> getModules() {
```
Comment (cstamas) [inline maven-core/src/main/java/org/apache/maven/project/MavenProject.java:710]: Deprecate it? Or maybe does not matter, as in API 4 plugins have no reach to this class?

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/ModelBuilder.java [hunk 2]
```diff
@@ -25,7 +25,11 @@
 
 public interface ModelBuilder extends Service {
 
-    List<String> VALID_MODEL_VERSIONS = List.of("4.0.0", "4.1.0");
+    String MODEL_VERSION_4_0_0 = "4.0.0";
+
+    String MODEL_VERSION_4_1_0 = "4.0.0";
```
Comment (desruisseaux) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/ModelBuilder.java:30]: Typo? Shouldn't the value be "4.1.0"?

**Comment by mthmulders** [comment]
Shouldn't the subprojects be transformed to modules in the consumer pom?

File: maven-core/src/main/java/org/apache/maven/project/MavenProject.java (hunk 1 above)
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/project/MavenProject.java:710]: Yes, I was assuming this change targets 4.x api and plugins, so I would not care about the v3 api.

**Comment by gnodet** [comment]
> Shouldn't the subprojects be transformed to modules in the consumer pom?

Actually, the modules/subprojects are build information, so not available in the consumer POM.

File: api/maven-api-core/src/main/java/org/apache/maven/api/plugin/annotations/Mojo.java [hunk 3]
```diff
@@ -76,8 +76,8 @@
     boolean projectRequired() default true;
 
     /**
-     * if the Mojo uses the Maven project and its child modules.
-     * @return uses the Maven project and its child modules
+     * if the Mojo uses the Maven project and its subprojects.
+     * @return uses the Maven project and its subprojectss
```
Comment (Bukama) [inline api/maven-api-core/src/main/java/org/apache/maven/api/plugin/annotations/Mojo.java:80]: `subprojectss` -> `subprojects`

File: maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/DefaultModelValidator.java [hunk 4]
```diff
@@ -360,6 +360,61 @@ public void validateFileModel(Model m, ModelBuilderRequest request, ModelProblem
...
+                                Severity.WARNING,
+                                Version.V41,
+                                "modules",
+                                null,
+                                "deprecated modules element, use subprojects instead",
+                                m.getLocation("modules"));
+                    } else {
+                        addViolation(
+                                problems,
+                                Severity.WARNING,
```
Comment (Bukama) [inline maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/DefaultModelValidator.java:409]: Why not error?

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 5]
```diff
@@ -121,6 +121,7 @@
 public class DefaultProjectBuilder implements ProjectBuilder {
     public static final String BUILDER_PARALLELISM = "maven.projectBuilder.parallelism";
     public static final int DEFAULT_BUILDER_PARALLELISM = Runtime.getRuntime().availableProcessors() / 2 + 1;
+    private static final String POM_4_0_0 = "4.0.0";
```
Comment (Bukama) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:124]: Why not `MODEL_VERSION_4.0.0` as above?

File: maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/DefaultModelValidator.java (hunk 4 above)
Comment (gnodet) [reply maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/DefaultModelValidator.java:409]: Good catch !

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java (hunk 5 above)
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:124]: That one isn't used anymore, I'll remove it.

**Review by cstamas — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #9793

**Pull request merged**